src/shell.html

```diff
@@ -3,6 +3,37 @@
   <head>
     <meta charset="utf-8">
     <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
+
+    <script>
+      var Module = {
+        preRun: [],
+        postRun: [],
+      };
+
+      // For best stability on 32-bit browsers, allocate asm.js/WebAssembly heap up front before proceeding
+      // to load any other page content. This mitigates the chances that loading up page assets first would
+      // fragment the memory area of the browser process.
+      var supportsWasm = (typeof WebAssembly === 'object' && typeof WebAssembly.Memory === 'function');
+      var page_size = supportsWasm ? 64 * 1024 : 16 * 1024 * 1024;
+      var TOTAL_MEMORY = page_size * Math.ceil({{{ TOTAL_MEMORY }}} / page_size);
+      var MAX_MEMORY = Math.max(TOTAL_MEMORY, page_size * Math.ceil({{{ BINARYEN_MEM_MAX }}} / page_size));
+
+      if (supportsWasm) {
+        Module['wasmMemory'] = new WebAssembly.Memory({ initial: TOTAL_MEMORY / page_size, maximum: MAX_MEMORY / page_size });
+        Module['buffer'] = Module['wasmMemory'].buffer;
+      } else { // asm.js:
+        for(var mem = MAX_MEMORY; mem >= TOTAL_MEMORY; mem -= multiple) {
+          try {
+            Module['buffer'] = new ArrayBuffer(mem);
+            if (Module['buffer'].byteLength != TOTAL_MEMORY) throw 'Out of memory';
```

Comment:
I want to provide the same mechanism that WebAssembly gives in the sense that one can ask to allocate way more (MAX_MEMORY), but is able to settle for a minimum required amount as well (TOTAL_MEMORY). This is something that has occasionally come up as useful.

Thought it to be nice and symmetric to reuse the same variable name that currently does this for Binaryen/WebAssembly as well. (although perhaps a bit odd then that it is named after Binaryen)